### PR TITLE
Follow friend

### PR DESCRIPTION
**Pull request by ramdog**
[#44734814360939]
[#31641682275895]

Open to suggestions on better ways to handle the query and processing, server-side.  It works, but definitely think there might be more elegant ways of achieving the same thing.  Basic idea:
  1.  Go get the user and the friend, if they don't both exist throw 404
  2.  Check that each user exists in each other's subdocument of `friends`.  If not, throw 404.
  3.  Update the `subscribeTo`, `subscribesToMe` properties.

image

### CODE REVIEW

File: server/api/friend/friend.controller.js [hunk 1]
```diff
@@ -4,35 +4,88 @@ var _ = require('lodash');
...
+  var followingUserId = req.user._id;
+  var followedUserId = req.params.id;
+  var subscribeTo = req.body.subscribeTo;
+
+  if (!_.isBoolean(req.body.subscribeTo)) {
+    // TODO: Does not seem like a great solution.  Would be good to think of a
+    // TODO: better validation solution as all the other routes are suffering
+    // TODO: from a similar problem(maybe some express middleware). Maybe
+    // TODO: there is some Express middleware.  I guess it's not a big problem,
+    // TODO: but remember this is an engineering exercise first!
```
Comment (guilload) [inline server/api/friend/friend.controller.js]: :)

Comment (ramdog) [inline server/api/friend/friend.controller.js]: Haha ;)

**Comment by guilload** [comment]
Funny how we have to deal with inconsistencies at application level when using NoSQL databases. In the relational world, it would have been easy to model this many-to-many relationship, use foreign keys and let the database take care of that for us.

Looks like MongoDB has transactions, maybe we should use them since we need to modify 2 objects when updating a subscription. It's probably overkill given the size of the current user base, but remember, it's an engineering project :D

Also, I'm not a fan of the variable names `subscribeTo` and `subscribesToMe`, I would prefer `subscribing` and `subscribers`.

**Comment by ramdog** [comment]
@guilload - Thanks for the feedback.

Yes, while having MongoDb part of this has been an interesting learning experience, I gravitate toward using a relational DB myself.  I agree with the transaction idea would be worth looking into - very good idea.  There are a few places in the app where different models are being updated as part of a single "process" and wrapping that in a transaction would be a great idea.  I'll look into it.

On the naming, I don't love it either, but I'm not sure those other names make sense either.  Remember that the user model has a friends property that is described like this:

```js
friends: [{
    user: {type: Schema.Types.ObjectId, ref: 'User'},
    subscribeTo: {type: Boolean, default: false},
    subscribesToMe: {type: Boolean, default: false},
    connectedOn: {type: Date, default: new Date()}
  }],
```

So, the `subscribTo` and `subscribesToMe` are booleans.  In theory, I actually didn't need to uses both - I could just store if the user subscribes to their friend or not.  Perhaps I'm already developing bad habits by using Mongo, but I figured the Mongo way would be to store that info in both places so I wouldn't have to go and query for every friend in a user's friends when preparing the notifications.  That said, it seems like a bad idea to replicate the data, doesn't it?  Hmm...

**Comment by guilload** [comment]
Right, I had totally missed that...

**Pull request merged**